src/main/java/dev/suvera/keycloak/scim2/storage/storage/ScimSyncRunner.java

```diff
@@ -20,84 +21,114 @@
 import dev.suvera.keycloak.scim2.storage.jpa.ScimSyncJobQueue;
 
 public class ScimSyncRunner {
+    private static final Logger log = Logger.getLogger(ScimSyncRunner.class);
     private KeycloakSessionFactory sessionFactory;
-    private KeycloakSession session;
     private ComponentModel model;
 
     public ScimSyncRunner(KeycloakSessionFactory sessionFactory, ComponentModel model) {
         this.sessionFactory = sessionFactory;
         this.model = model;
-        session = sessionFactory.create();
     }
 
     public SynchronizationResult syncAll(String realmId) {
+        log.infof("Starting synchronization of all users for realm: %s", realmId);
+
         SynchronizationResult result = new SynchronizationResult();
 
-        callSyncJobs(session, result);
+        List<String> ldapComponentModels = ComponentModelUtils
+                .getLDAPComponentsWithScimEventsEnabled(sessionFactory, realmId)
+                .collect(Collectors.toList());
 
-        KeycloakModelUtils.runJobInTransaction(sessionFactory, kcSession -> {
-            kcSession.getContext().setRealm(kcSession.realms().getRealm(realmId));
+        KeycloakSession session = sessionFactory.create();
+        RealmModel realm = session.realms().getRealm(realmId);
+
+        int batchNumber = 0;
+        List<UserModel> users;
+        do {
+            log.infof("Fetching batch %d of users...", batchNumber+1);
+            users = getBatchOfUsers(realm, ldapComponentModels, 1000, batchNumber);
+            log.infof("Fetched %d users in batch %d.", users.size(), batchNumber+1);
+
+            users.forEach(user -> executeSyncUserJob(user, realmId, result));
+            log.infof("Executed synchronization of users for batch %d.", batchNumber+1);
+
+            batchNumber++;
```

Comment:
```suggestion
            log.infof("Fetching batch %d of users...", ++batchNumber);
            users = getBatchOfUsers(realm, ldapComponentModels, 1000, batchNumber-1);
            log.infof("Fetched %d users in batch %d.", users.size(), batchNumber);

            users.forEach(user -> executeSyncUserJob(user, realmId, result));
            log.infof("Executed synchronization of users for batch %d.", batchNumber);
```

Less operations. Alternatively, and even better IMO is to just log the "true" batch number and increment in the end, as you did. I see no particular reason to be obliged to log things with 1-based indexes when in reality it's 0-based.

Reply:
Agreed. Opted not incrementing the batch number for logging.